File: src/git-secret/report.py
import enum
from collections import namedtuple
from arguments import Arguments


class ReportCode(enum.Enum):
    SUCCESS = 0
    UNSUPPORTED = 1
    FAILED = 2


ReportSeverityLevelTupe = namedtuple('SeverityLevel',
                                     ['code', 'short'])
class ReportSeverityLevel(enum.Enum):
    INFO = ReportSeverityLevelTupe(0, 'I')
    WARN = ReportSeverityLevelTupe(1, 'W')
    ERROR = ReportSeverityLevelTupe(2, 'E')

    @property
    def code(self):
        return self.value.code

    @property
    def short(self):
        return self.value.short


ReportIncidentCodeTupe = namedtuple('IncidentCode',
                                           ['code',
                                            'description'])
class ReportIncidentCode(enum.Enum):
    BLACKLIST_FILENAME = ReportIncidentCodeTupe('B001', 'Blacklisted filename')
    BLACKLIST_STRING = ReportIncidentCodeTupe('E001', 'Blacklisted string')
    ENTROPY_STRING = ReportIncidentCodeTupe('E002', 'High entropy string')

    @property
    def code(self):
        return self.value.code

    @property
    def description(self):
        return self.value.description


class ReportIncident():
    Code = ReportIncidentCode

    def __init__(self, serverity, code, filename: str, offend: str, author: str):
        self.serverity = serverity
        self.code = code
        self.filename = filename
        self.offend = offend
        self.author = author

    def dump(self, verbose=True):
        if not verbose and (self.serverity == ReportSeverityLevel.INFO or self.serverity == ReportSeverityLevel.WARN):
            return
        print("- [{serverity}:{code}] {message}\n"
              "           File  : {file}\n"
              "           Author: {author}".format(
                serverity=self.serverity.short,
                code=self.code.code,
                file=self.filename,
                offend=self.offend,
                author=self.author,
                message=self.code.description))
        if self.serverity == Report.SeverityLevel.ERROR and self.offend:
            print("         * Offend: {}".format(self.offend))


class Report:
    Code = ReportCode
    SeverityLevel = ReportSeverityLevel
    Incident = ReportIncident

    def __init__(self, code=ReportCode.FAILED):
        self.code = code
        self.incidents = []

    def dump(self, verbose=True):
        print("[ ] Scan result: {}".format(self.code.name))

        reordered = sorted(self.incidents, key=lambda x: x.serverity.code, reverse=True)
        for incident in reordered:
            incident.dump(verbose)

File: src/ignorefile/report/report.ml
type severity_level = Info | Warn | Error

type incident_code =
  | Blacklisted_filename
  | Blacklisted_string
  | Entropy_string
  | Missing_ignore_pattern of string

type incident =
  {filename: string; code: incident_code; severity_level: severity_level}

let int_of_severity_level = function Info -> 0 | Warn -> 1 | Error -> 2

let short_string_of_severity_level = function
  | Info ->
      "I"
  | Warn ->
      "W"
  | Error ->
      "E"

let long_string_of_severity_level = function
  | Info ->
      "INFO"
  | Warn ->
      "WARN"
  | Error ->
      "ERROR"

let code_of_incident_code = function
  | Blacklisted_filename ->
      "B001"
  | Blacklisted_string ->
      "E001"
  | Entropy_string ->
      "E002"
  | Missing_ignore_pattern _ ->
      "E003"

let message_of_incident_code = function
  | Blacklisted_filename ->
      "Blacklisted filename"
  | Blacklisted_string ->
      "Blacklisted string"
  | Entropy_string ->
      "High entropy string"
  | Missing_ignore_pattern s ->
      "Missing ignore pattern `" ^ s ^ "'"

let pp_incident ppf incident =
  let open Format in
  fprintf ppf "@[<v>@[<h>- [%a:%a] %a@]@,           File  : %a@]"
    pp_print_string
    (short_string_of_severity_level incident.severity_level)
    pp_print_string
    (code_of_incident_code incident.code)
    pp_print_string
    (message_of_incident_code incident.code)
    pp_print_string incident.filename

let pp_incidents ppf incidents = Format.pp_print_list pp_incident ppf incidents
